File: days/src/main/java/aoc2024j/day4/Day4.java

package aoc2024j.day4;

import utils.IO;

import java.util.List;

public class Day4 {

    long part1(List<String> data) {
        var grid = new Part1Grid(data, "XMAS");
        return grid.count();
    }

    long part2(List<String> data) {
        var grid = new Part2Grid(data, "MAS");
        return grid.count();
    }

    public static void main(String[] args) {
        var day4 = new Day4();
        var data = IO.getResourceAsList("aoc2024/day4.txt");
        var part1 = day4.part1(data);
        System.out.println("part1 = " + part1);
        var part2 = day4.part2(data);
        System.out.println("part2 = " + part2);
    }
}

File: days/src/main/scala/aoc2024s/day4/Day4.scala

package aoc2024s.day4;

import utils.IO;
import scala.jdk.CollectionConverters.*

object Day4 {

  def part1(data: List[String]): Long = {
    Part1Grid(data).count("XMAS")
  }

  def part2(data: List[String]): Long = {
    Part2Grid(data).count("MAS");
  }

  @main def main4(): Unit = {
    val data = IO.getResourceAsList("aoc2024/day4.txt").asScala.toList;
    val part1 = Day4.part1(data);
    println(s"part1 = $part1");
    val part2 = Day4.part2(data);
    println(s"part2 = $part2");
  }
}